File: lab3/java/SearchResult.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * The class for search results.
 * 
 * If you found a result: `SearchResult(graph, True, start, goal, cost, path, iterations)`
 * If no path was found: `SearchResult(graph, False, start, goal, -1, None, iterations)`
 */
public class SearchResult<V> {
    public final Graph<V> graph;
    public final boolean success;
    public final V start;
    public final V goal;
    public final double cost;
    public final List<Edge<V>> path;
    public final int iterations;

    SearchResult(Graph<V> graph, boolean success, V start, V goal, double cost, List<Edge<V>> path, int iterations) {
        this.graph = graph;
        this.success = success;
        this.start = start;
        this.goal = goal;
        this.cost = cost;
        this.path = path;
        this.iterations = iterations;
    }

    private String formatPathPart(boolean suffix, boolean withWeight, int i, int j) {
        return path.subList(i, j).stream()
                .map(e -> e.toString(!suffix, suffix, withWeight && this.graph.isWeighted()))
                .collect(Collectors.joining());
    }

    public void validate() {
        if (this.success) {
            if (this.path == null)
                throw new IllegalArgumentException("The path must be a list - did you forget to implement extractPath?");
            double actualCost = this.path.stream()
                .mapToDouble((e) -> e.weight)
                // We sum using left association order to mimic the algorithm.
                .reduce(0, Double::sum);
            if (this.cost != actualCost) 
                throw new IllegalArgumentException(String.format(
                    "The reported path cost (%f) differs from the calculated cost (%f)", this.cost, actualCost
                ));
        }
        if (this.iterations <= 0)
            throw new IllegalArgumentException("The number of iterations should be > 0");
    }

    public String toString(boolean drawGraph, int maxGraphWidth, int maxGraphHeight, boolean withWeight) {
        ArrayList<String> lines = new ArrayList<>();
        if (this.iterations <= 0)
            lines.add("ERROR: You have to iterate over at least the starting state!");
        lines.add("Loop iterations: " + this.iterations);
        if (drawGraph) {
            String graphStr = this.graph.drawGraph(maxGraphWidth, maxGraphHeight, this.start, this.goal, this.path);
            if (graphStr != null)
                lines.add(graphStr + "\n");
        }
        if (!this.success) {
            lines.add(String.format("No path found from %s to %s", start, goal));
        } else {
            lines.add(String.format("Cost of path from %s to %s: %.2f", start, goal, cost));
            if (path == null) {
                lines.add("WARNING: You have not implemented extractPath!");
            } else {
                // Print the path.
                int pathlen = this.path.size();
                lines.add("Number of edges: " + pathlen);
                if (pathlen <= 10) {
                    lines.add(
                        this.path.get(0).start + 
                        formatPathPart(true, withWeight, 0, pathlen)
                    );
                } else {
                    lines.add(
                        formatPathPart(false, withWeight, 0, 5) + "..." + 
                        formatPathPart(true, withWeight, pathlen-5, pathlen)
                    );
                }
                double actualCost = this.path.stream()
                    .mapToDouble((e) -> e.weight)
                    // We sum using left association order to mimic the algorithm.
                    .reduce(0, Double::sum);
                if (this.cost != actualCost) {
                    lines.add(String.format(
                        "WARNING: The actual path cost (%f) differs from the reported cost (%f)!", 
                        actualCost, this.cost
                    ));
                }
            }
        }
        return String.join("\n", lines);
    }

}

File: lab3/python/search_result.py
from dataclasses import dataclass
from typing import Generic, List, Optional

from edge import Edge, V
from graph import Graph


# Dataclasses provide a lot of methods for free, such as __init__ and comparison:
# https://docs.python.org/3/library/dataclasses.html
# 'frozen' means that we cannot modify elements after creation:
@dataclass(frozen=True)
class SearchResult(Generic[V]):
    """
    A class for search results.

    If you found a result: `SearchResult(graph, True, start, goal, cost, path, iterations)`
    If no path was found: `SearchResult(graph, False, start, goal, -1, None, iterations)`
    """
    graph: Graph[V]
    success: bool
    start: V
    goal: V
    cost: float
    path: Optional[List[Edge[V]]]
    iterations: int

    def formatPathPart(self, suffix: bool, withWeight: bool, i: int, j: int) -> str:
        assert self.path  # The path must not be None
        return "".join(
            e.toString(not suffix, suffix, withWeight and self.graph.isWeighted())
            for e in self.path[i:j]
        )

    def validate(self):
        if self.success:
            if not isinstance(self.path, list):
                raise ValueError("The path must be a list - did you forget to implement extractPath?")
            actualCost = sum(e.weight for e in self.path)
            if (self.cost != actualCost):
                raise ValueError("The reported path cost ({self.cost}) differs from the calculated cost ({actualCost})")
        if self.iterations <= 0:
            raise ValueError("The number of iterations should be > 0")

    def toString(self, 
                drawGraph: bool,
                maxGraphWidth: int,
                maxGraphHeight: int,
                withWeight: bool,
            ) -> str:
        lines: List[str] = []
        if self.iterations <= 0:
            lines.append("ERROR: You have to iterate over at least the starting state!")
        lines.append(f"Loop iterations: {self.iterations}")
        if drawGraph:
            graphStr = self.graph.drawGraph(maxGraphWidth, maxGraphHeight, self.start, self.goal, self.path)
            if graphStr:
                lines.append(graphStr + "\n")
        if not self.success:
            lines.append(f"No path found from {self.start} to {self.goal}")
        else:
            lines.append(f"Cost of path from {self.start} to {self.goal}: {self.cost:.2f}")
            if not isinstance(self.path, list):
                lines.append("WARNING: You have not implemented extractPath!")
            else:
                # Print the path.
                pathlen = len(self.path)
                lines.append(f"Number of edges: {pathlen}")
                if pathlen <= 10:
                    lines.append(
                        str(self.path[0].start) + 
                        self.formatPathPart(True, withWeight, 0, pathlen)
                    )
                else:
                    lines.append(
                        self.formatPathPart(False, withWeight, 0, 5) + "..." +
                        self.formatPathPart(True, withWeight, pathlen-5, pathlen)
                    )
                actualCost = sum(e.weight for e in self.path)
                if self.cost != actualCost:
                    lines.append(
                        f"WARNING: the actual path cost ({actualCost}) "
                        f"differs from the reported cost ({self.cost})!"
                    )
        return '\n'.join(lines)


